### PR TITLE
feat: Added Paymaster Proxy

### PR DESCRIPTION
**Pull request by cpcramer**
**What changed? Why?**
Add paymaster proxy.
- Add paymaster proxy route
- Add CoinbaseSmartWallet ABI
- Add unit tests for the route.

**Notes to reviewers**

**How has it been tested?**

### CODE REVIEW

File: template/web/app/paymaster-bundler/api/paymasterProxy/route.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { NextRequest, NextResponse } from 'next/server';
+import { UserOperation } from 'permissionless';
+import { paymasterClient } from '../../../../../web/src/components/SmartWallets/PaymasterClient';
+
+type PaymasterRequest = {
+  method: string;
+  params: [UserOperation<'v0.6'>, string, string];
+};
+
```
Comment (Zizzamia) [inline template/web/app/paymaster-bundler/api/paymasterProxy/route.ts:9]: Reflecting on the name of the file `/api/paymasterProxy`, it means the api will be CamelCase, which I don't think it's a good practice.

should this be `paymaster-proxy`?

Comment (cpcramer) [reply template/web/app/paymaster-bundler/api/paymasterProxy/route.ts:9]: Hmm perhaps, other apis in the repo follow camelCase though https://github.com/coinbase/build-onchain-apps/blob/paul/add-paymaster-proxy/template/web/app/api/chain/currentBlockNumber/route.ts

Comment (cpcramer) [reply template/web/app/paymaster-bundler/api/paymasterProxy/route.ts:9]: I feel think `paymaster-proxy` is better. I can refactor all of the routes in this repo as a fast follow

File: template/web/app/paymaster-bundler/api/paymasterProxy/route.test.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import { NextRequest } from 'next/server';
+import { paymasterClient } from '../../../../../web/src/components/SmartWallets/PaymasterClient';
```
Comment (Zizzamia) [inline template/web/app/paymaster-bundler/api/paymasterProxy/route.test.ts:2]: PaymasterClient feels more an utils, something that should live under:
- https://github.com/coinbase/build-onchain-apps/tree/main/template/web/src/utils/paymasterClient.ts

and than accessed as `@/utils/paymasterClient`

File: template/web/app/paymaster-bundler/api/paymasterProxy/route.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * @returns {Promise<Response>} - The response object containing the result or an error message.
+ */
+export async function POST(req: NextRequest): Promise<NextResponse> {
+  const reqBody: PaymasterRequest = (await req.json()) as PaymasterRequest;
+  const { method, params } = reqBody;
+  const [userOp] = params;
+
+  try {
+    let result;
+    if (method === 'pm_getPaymasterStubData') {
```
Comment (Zizzamia) [inline template/web/app/paymaster-bundler/api/paymasterProxy/route.ts:23]: `pm_getPaymasterStubData` and `pm_getPaymasterData` could be standardized into TS ENUM

Comment (Zizzamia) [reply template/web/app/paymaster-bundler/api/paymasterProxy/route.ts:23]: Also what's the diffference between those statements? like for someone not expert with Paymaster how will you explain those if/else?

File: template/web/package.json [hunk 4]
```diff
@@ -21,7 +21,6 @@
     "@coinbase/onchainkit": "0.16.10",
     "@coinbase/wallet-sdk": "4.0.0",
     "@farcaster/hub-nodejs": "^0.11.9",
-    "@privy-io/react-auth": "^1.56.2",
```
Comment (Zizzamia) [inline template/web/package.json:24]: nice

File: template/web/src/components/SmartWallets/PaymasterClient.ts [hunk 5]
```diff
@@ -1,7 +1,16 @@
-import { createPublicClient, http } from 'viem';
+import { ENTRYPOINT_ADDRESS_V06 } from 'permissionless';
+import { paymasterActionsEip7677 } from 'permissionless/experimental';
+import { createPublicClient, createClient, http } from 'viem';
 import { baseSepolia } from 'viem/chains';
 
+const paymasterService = process.env.NEXT_PUBLIC_PAYMASTER_URL ?? '';
+
 export const client = createPublicClient({
```
Comment (Zizzamia) [inline template/web/src/components/SmartWallets/PaymasterClient.ts:8]: As this is a general util can just live in `src/utils/client` and `src/utils/paymasterClient`

File: template/web/app/paymaster-bundler/api/paymasterProxy/route.ts (hunk 3 above)
Comment (cpcramer) [reply template/web/app/paymaster-bundler/api/paymasterProxy/route.ts:23]: afaik `pm_getPaymasterStubData` is the initial call which does a quick check/validation to see if the transaction is sponsorable. 

If things look ok then `pm_getPaymasterData` is called with real (as opposed to stub) data and this is where sponsorship actually happens. 

I'm adding a small comment above each method for clarity

File: template/web/app/paymaster-bundler/api/paymasterProxy/route.test.ts (hunk 2 above)
Comment (cpcramer) [reply template/web/app/paymaster-bundler/api/paymasterProxy/route.test.ts:2]: moved!

File: template/web/src/components/SmartWallets/PaymasterClient.ts (hunk 5 above)
Comment (cpcramer) [reply template/web/src/components/SmartWallets/PaymasterClient.ts]: I moved the paymasterClient to be under `@/utils/`

File: template/web/app/paymaster-bundler/_contracts/NewPaymasterBundlerContract.test.ts [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import { baseSepolia } from 'viem/chains';
 import { SMART_WALLET_ENTRY_POINT_ADDRESS } from '../constants';
-import { abi } from './NewPaymasterBundlerABI';
-import { useNewPaymasterBundlerContract } from './useNewPaymasterBundlerContract';
+import { useNewPaymasterBundlerContract } from './NewPaymasterBundlerContract';
+import { abi } from './PaymasterBundlerABI';
```
Comment (cpcramer) [inline template/web/app/paymaster-bundler/_contracts/NewPaymasterBundlerContract.test.ts:4]: Will update and remove the `new`

File: template/web/app/paymaster-bundler/api/paymasterProxy/route.ts (hunk 1 above)
Comment (Zizzamia) [reply template/web/app/paymaster-bundler/api/paymasterProxy/route.ts:9]: nice

File: template/web/app/paymaster-bundler/api/paymasterProxy/route.ts (hunk 3 above)
Comment (Zizzamia) [reply template/web/app/paymaster-bundler/api/paymasterProxy/route.ts:24]: nice

**Review by Zizzamia — APPROVED** [review]

**Pull request merged**